samples/sample_decode/src/pipeline_decode.cpp

```diff
@@ -1884,6 +1904,36 @@ mfxStatus CDecodingPipeline::RunDecoding()
                 bErrIncompatibleVideoParams = true;
                 // need to go to the buffering loop prior to reset procedure
                 pBitstream = NULL;
+                sts = MFX_ERR_NONE;
+                continue;
+            } else if (MFX_ERR_REALLOC_SURFACE == sts) {
+                mfxVideoParam param;
+                m_pmfxDEC->GetVideoParam(&param);
+
+                m_pCurrentFreeSurface->frame.Info.CropW = param.mfx.FrameInfo.CropW;
+                m_pCurrentFreeSurface->frame.Info.CropH = param.mfx.FrameInfo.CropH;
+                m_YUV_Width = MSDK_ALIGN16(std::max(param.mfx.FrameInfo.Width, (mfxU16)m_YUV_Width));
+                m_YUV_Height = MSDK_ALIGN16(std::max(param.mfx.FrameInfo.Height, (mfxU16)m_YUV_Height));
+                m_pCurrentFreeSurface->frame.Info.Width = m_YUV_Width;
+                m_pCurrentFreeSurface->frame.Info.Height = m_YUV_Height;
+
+                inMid = m_pCurrentFreeSurface->frame.Data.MemId;
+                ReallocSurface(inMid, &m_pCurrentFreeSurface->frame.Info, m_pCurrentFreeSurface->frame.Data.MemType, &outMid);
+
+                // We have to lock surface in case of system memory allocator to have right pitches in Data
+                // (that can be needed for INTERNAL->SYS copy operations)
+                // Better way is locking surface in decoder, but we don't know real memid inside decoders (in case of sysmem)
+                // It's a good point for improvement
+                if (m_pCurrentFreeSurface->frame.Data.MemType & MFX_MEMTYPE_SYSTEM_MEMORY)
+                {
+                    m_pGeneralAllocator->Lock(m_pGeneralAllocator->pthis, outMid, &(m_pCurrentFreeSurface->frame.Data));
+                    m_pGeneralAllocator->Unlock(m_pGeneralAllocator->pthis, outMid, NULL);
```

Comment:
Looks weird. You carry out information out of lock/unlock scope. I feel this should be done in some other way.

Reply:
It is necessary to calculate the offset of Y/U/V address. If we have better solution, I can change it. Thanks.

Reply:
I still don't see why this is needed. If you use external allocator (even with system memory), then you _must_ call Lock before using the memory and Unlock once you are done. This is what mediasdk library should do. If it is doing something else and tries directly access system memory - it's just a bug in mediasdk.

Can you, please, remove this code and provide details where you will see the crash?

Reply:
I will check it. Thanks

Reply:
CDecodingPipeline::AllocFrames() will call LockFrame to calculate UV offset and pitch after create surface.
So after reallocate the surface, we also need call LockFrame() to calculate them. 
May it be better to move this calling into CDecodingPipeline::ReallocSurface()? If yes, I can change it.

Reply:
I also check the possibility to call LockFrame in Decoder. But VideoDECODEVP9_HW::m_core only keep internal memory ID and corresponding allocator reference. So it seems that it cannot handle the MemId from sample_decode.

Reply:
Attach the full back trace when crash for reference:
Thread 2 "sample_decode" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff515e700 (LWP 18140)]
0x00007ffff54119dd in _mm_storeu_si128(long long __vector(2)*, long long __vector(2)) (__B=..., __P=0x7ffff7fd7000) at /usr/lib/gcc/x86_64-redhat-linux/8/include/emmintrin.h:720
720	  *__P = __B;
Missing separate debuginfos, use: dnf debuginfo-install libX11-1.6.5-7.fc28.x86_64 libXau-1.0.8-11.fc28.x86_64 libXext-1.3.3-8.fc28.x86_64 libXfixes-5.0.3-5.fc28.x86_64 libgcc-8.2.1-4.fc28.x86_64 libpciaccess-0.13.4-8.fc28.x86_64 libstdc++-8.1.1-5.fc28.x86_64 libxcb-1.13-1.fc28.x86_64
(gdb) bt
#0  0x00007ffff54119dd in _mm_storeu_si128(long long __vector(2)*, long long __vector(2)) (__B=..., __P=0x7ffff7fd7000) at /usr/lib/gcc/x86_64-redhat-linux/8/include/emmintrin.h:720
#1  copyVideoToSys (src=0x7ffff7fb9480 <error: Cannot access memory at address 0x7ffff7fb9480>, dst=0x7ffff7fd6fe0 "", width=544)
    at /home/yanwang/open_media_workspace/MediaSDK/_studio/shared/include/fast_copy.h:64
#2  0x00007ffff5412c1d in mfxCopyRect<unsigned char> (pSrc=0x7ffff7fb9480 <error: Cannot access memory at address 0x7ffff7fb9480>, srcStep=640, pDst=0x7ffff7fd6fe0 "", dstStep=576, roiSize=..., flag=2)
    at /home/yanwang/open_media_workspace/MediaSDK/_studio/shared/include/fast_copy.h:214
#3  0x00007ffff54123e9 in FastCopy::Copy (pDst=0x7ffff7fb90a0 <error: Cannot access memory at address 0x7ffff7fb90a0>, dstPitch=576, pSrc=0x7ffff7f98000 <error: Cannot access memory at address 0x7ffff7f98000>, 
    srcPitch=640, roi=..., flag=2) at /home/yanwang/open_media_workspace/MediaSDK/_studio/shared/include/fast_copy.h:250
#4  0x00007ffff540fa65 in CoreDoSWFastCopy (pDst=0x7ffff515d720, pSrc=0x7ffff515d7e0, copyFlag=2) at /home/yanwang/open_media_workspace/MediaSDK/_studio/shared/src/libmfx_core.cpp:1321
#5  0x00007ffff541e653 in VAAPIVideoCORE::DoFastCopyExtended (this=0x5555557ec470, pDst=0x7ffff515d720, pSrc=0x7ffff515d7e0)
    at /home/yanwang/open_media_workspace/MediaSDK/_studio/shared/src/libmfx_core_vaapi.cpp:1178
#6  0x00007ffff541e010 in VAAPIVideoCORE::DoFastCopyWrapper (this=0x5555557ec470, pDst=0x555555826bc0, dstMemType=66, pSrc=0x7ffff515d9a0, srcMemType=17)
    at /home/yanwang/open_media_workspace/MediaSDK/_studio/shared/src/libmfx_core_vaapi.cpp:1032
#7  0x00007ffff542628e in mfx_UMC_FrameAllocator_D3D::PrepareToOutput (this=0x55555581c8c0, surface_work=0x555555826bc0, index=2, isOpaq=false)
    at /home/yanwang/open_media_workspace/MediaSDK/_studio/shared/src/mfx_umc_alloc_wrapper.cpp:1108
#8  0x00007ffff5501ed2 in VP9DECODERoutine (p_state=0x555555851000) at /home/yanwang/open_media_workspace/MediaSDK/_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp:852
#9  0x00007ffff53ac94c in mfxSchedulerCore::call_pRoutine (this=0x5555557ec940, call=...) at /home/yanwang/open_media_workspace/MediaSDK/_studio/mfx_lib/scheduler/include/mfx_scheduler_core.h:500
#10 0x00007ffff53ac687 in mfxSchedulerCore::ThreadProc (this=0x5555557ec940, pContext=0x555555806c88)
    at /home/yanwang/open_media_workspace/MediaSDK/_studio/mfx_lib/scheduler/src/mfx_scheduler_core_thread.cpp:80

Reply:
Ok, so, if I read code correctly you actually should have called LockFrame here:

https://github.com/Intel-Media-SDK/MediaSDK/blob/45b78442f7da338e69b2774cd448047c1c8b2055/_studio/shared/src/libmfx_core_vaapi.cpp#L987

Can you, please, check:
1) Was it actually called or not? If not - we likely use cached pointers from the previous allocation.
2) If it was called, did we somehow get pointers from prev. allocation? I am not sure why this could happen though....

Reply:
Sure. I will check it. Thanks.

Reply:
I have investigated it and updated this PR. About your questions, I add my comments in the following: 

> Ok, so, if I read code correctly you actually should have called LockFrame here:
> 
> https://github.com/Intel-Media-SDK/MediaSDK/blob/45b78442f7da338e69b2774cd448047c1c8b2055/_studio/shared/src/libmfx_core_vaapi.cpp#L987
> 
> Can you, please, check:
> 
> 1. Was it actually called or not? If not - we likely use cached pointers from the previous allocation.
Currently it isn't called because of dstPtr != nullptr. CDecodingPipeline::AllocFrames() will call LockFrame after create surface.
> 2. If it was called, did we somehow get pointers from prev. allocation? I am not sure why this could happen though....
After I removed LockFrame in CDecodingPipeline::AllocFrames(), this can be called but no available external frame allocator because sample_decode will not set external frame allocator when using system memory in CDecodingPipeline::CreateAllocator(). 
https://github.com/Intel-Media-SDK/MediaSDK/blob/45b78442f7da338e69b2774cd448047c1c8b2055/samples/sample_decode/src/pipeline_decode.cpp#L1386.
After I added it in the latest update of this PR, this issue is fixed.

Reply:
Currently the only possible issue is that VideoDECODEVP9_HW::Init() will try to create internal frames by external frame allocator (system memory allocator) but failed because it wants to create MFX_MEMTYPE_DXVA2_DECODER_TARGET surface for VAAPI acceleration. After it get MFX_ERR_UNSUPPORTED from system memory allocator, it will fall back to use default allocator in the following line. So it may be OK.
https://github.com/Intel-Media-SDK/MediaSDK/blob/45b78442f7da338e69b2774cd448047c1c8b2055/_studio/shared/src/libmfx_core_vaapi.cpp#L584